src/firmware.cpp

```diff
@@ -192,109 +160,100 @@ static CommandError ultrasound_read(int commandId, String argument) {
+  if (pin < 2 || pin > 13) {
+    return COMMAND_ERROR("pin must be between 2 and 13");
+  }
+
+  switch(pinStateArg.charAt(0)){
+    case 'H':
+      pinMode(pin, OUTPUT);
+      digitalWrite(pin, HIGH);
+      break;
+    case 'L':
+      pinMode(pin, OUTPUT);
+      digitalWrite(pin, LOW);
+      break;
+    case 'P':
+      pinMode(pin, INPUT_PULLUP);
+      break;
+    case 'Z':
+      pinMode(pin, INPUT); // High Impedance
+      break;
+    default:
+      return COMMAND_ERROR("Unknown pin mode");
+    
+  }
+  return OK;
 }
 
-static void dispatch_command(int commandId, const class CommandHandler& handler, const String& argument) {
-  auto err = handler.run(commandId, argument);
+// Process the commands and execute them.
+
+static const CommandHandler commands[] = {
+  CommandHandler('A', &analogueRead), // Read the analogue pins
+  CommandHandler('L', &led), // Control the debug LED (H/L)
+  CommandHandler('R', &readPin), // Read a digital pin <number>
+  CommandHandler('S', &servo), // Control a servo <num> <width>
+  CommandHandler('U', & ultrasoundRead), // Read an ultrasound distance
+  CommandHandler('V', &version), // Get firmware version
+  CommandHandler('W', &writePin), // Write to or  a GPIO pin <number> <state>
+};
+
+static void dispatch_command(int requestID, const class CommandHandler& handler, const String& argument) {
+  auto err = handler.run(requestID, argument);
   if (err == OK) {
-    serialWrite(commandId, '+', "OK");
+    serialWrite(requestID, '+', "OK");
   } else {
-    serialWrite(commandId, '-', String("Error: ") + err);
+    serialWrite(requestID, '-', "Error: " + err);
   }
 }
 
-static void handle_actual_command(int commandId, const String& cmd) {
+static void handle_actual_command(int requestID, const String& cmd) {
+  
+    char commandIssued = cmd.charAt(0);
+    
     for (int i = 0; i < sizeof(commands) / sizeof(CommandHandler); ++i) {
       const CommandHandler& handler = commands[i];
 
-      if (handler.command == cmd) {
-        dispatch_command(commandId, handler, "");
-        return;
-      } else if (cmd.startsWith(handler.command + " ")) {
-        dispatch_command(
-          commandId,
-          handler,
-          cmd.substring(handler.command.length() + 1)
-        );
+      if (handler.command == commandIssued) {
+        dispatch_command(requestID, handler, cmd.substring(1));
         return;
       }
     }
 
-    serialWrite(commandId, '-', String("Error, unknown command: ") + cmd);
+    serialWrite(requestID, '-', String("Error, unknown command: ") + commandIssued);
 }
 
 static void handle_command(const String& cmd) {
+  // A command is prefixed with @ if an identifier is used to prevent race conditions
   if (cmd.startsWith("@")) {
     auto spaceIndex = cmd.indexOf(' ');
-    auto commandId = cmd.substring(1, spaceIndex).toInt();
-    handle_actual_command(commandId, cmd.substring(spaceIndex + 1));
+    auto requestID = cmd.substring(1, spaceIndex).toInt();
```

Comment:
Incoming point of argument: This should probably be `requestId`

Reply:
I'm unsure about what you're pointing out here. Please could you rephrase it?

Reply:
This was a nitpick spelling point. The variable should be called `requestId` rather than `requestID`.